File: src/Packet.Ax25/Session/GuardEvaluator.cs
namespace Packet.Ax25.Session;

/// <summary>
/// Evaluates the small boolean expression language used in SDL transition
/// <c>guard:</c> fields. Grammar:
/// <code>
///   expr   := term ("or" term)*
///   term   := factor ("and" factor)*
///   factor := "not"? identifier
/// </code>
/// Identifiers are arbitrary names resolved against an externally-supplied
/// binding table (e.g. <c>"own_receiver_busy"</c> → a lambda reading
/// <c>ctx.OwnReceiverBusy</c>).
/// </summary>
/// <remarks>
/// Whitespace separates tokens. Parentheses are not supported — every
/// guard observed so far in the spec is a simple conjunction of negated
/// or unnegated flags, occasionally an <c>or</c>. We extend the grammar
/// only as new SDL pages need it.
/// </remarks>
public sealed class GuardEvaluator
{
    private static readonly char[] TokenSeparators = { ' ', '\t' };

    private readonly IReadOnlyDictionary<string, Func<bool>> bindings;

    /// <summary>
    /// Build an evaluator over <paramref name="bindings"/>. Each entry maps
    /// an identifier name to a thunk that returns the identifier's current
    /// boolean value. Thunks are evaluated at each <see cref="Evaluate"/>
    /// call, so capturing mutable state via closure is the normal pattern.
    /// </summary>
    public GuardEvaluator(IReadOnlyDictionary<string, Func<bool>> bindings)
    {
        this.bindings = bindings ?? throw new ArgumentNullException(nameof(bindings));
    }

    /// <summary>
    /// Evaluate <paramref name="expression"/>. Returns <c>true</c> if the
    /// guard holds. Empty / null / whitespace-only expression is treated as
    /// <c>true</c> (no guard).
    /// </summary>
    /// <exception cref="GuardEvaluationException">
    /// Thrown when the expression references an unbound identifier, or has
    /// a syntax error.
    /// </exception>
    public bool Evaluate(string? expression)
    {
        if (string.IsNullOrWhiteSpace(expression))
        {
            return true;
        }

        var tokens = expression.Split(TokenSeparators, StringSplitOptions.RemoveEmptyEntries);
        int index = 0;
        bool result = ParseOr(tokens, ref index, expression);
        if (index != tokens.Length)
        {
            throw new GuardEvaluationException($"trailing tokens in guard expression '{expression}' at position {index}");
        }
        return result;
    }

    private bool ParseOr(string[] tokens, ref int idx, string original)
    {
        var result = ParseAnd(tokens, ref idx, original);
        while (idx < tokens.Length && tokens[idx] == "or")
        {
            idx++;
            var right = ParseAnd(tokens, ref idx, original);
            result = result || right;
        }
        return result;
    }

    private bool ParseAnd(string[] tokens, ref int idx, string original)
    {
        var result = ParseFactor(tokens, ref idx, original);
        while (idx < tokens.Length && tokens[idx] == "and")
        {
            idx++;
            var right = ParseFactor(tokens, ref idx, original);
            result = result && right;
        }
        return result;
    }

    private bool ParseFactor(string[] tokens, ref int idx, string original)
    {
        if (idx >= tokens.Length)
        {
            throw new GuardEvaluationException($"expected identifier in '{original}' at position {idx}");
        }
        bool negate = false;
        if (tokens[idx] == "not")
        {
            negate = true;
            idx++;
            if (idx >= tokens.Length)
            {
                throw new GuardEvaluationException($"expected identifier after 'not' in '{original}'");
            }
        }
        var ident = tokens[idx++];
        if (!bindings.TryGetValue(ident, out var pred))
        {
            throw new GuardEvaluationException($"unbound identifier '{ident}' in '{original}' — add a binding before evaluating this guard");
        }
        var value = pred();
        return negate ? !value : value;
    }
}

/// <summary>Thrown when a guard expression is malformed or references unbound identifiers.</summary>
public sealed class GuardEvaluationException : Exception
{
    public GuardEvaluationException(string message) : base(message) { }
}

File: web/ax25/src/sdl/guard-evaluator.ts
/**
 * Evaluates the small boolean expression language used in SDL transition
 * `guard:` fields. Grammar:
 *
 *   expr   := term ("or" term)*
 *   term   := factor ("and" factor)*
 *   factor := "not"? identifier
 *
 * Identifiers are resolved against an externally-supplied bindings table
 * (e.g. `"own_receiver_busy"` → a closure reading `ctx.ownReceiverBusy`).
 * Whitespace separates tokens. Parentheses are not supported — every guard
 * observed in the spec is a simple conjunction of negated or unnegated
 * flags, occasionally an `or`. The grammar grows only as new SDL pages
 * need it.
 *
 * Empty / null / whitespace-only expression is treated as `true` (no guard).
 *
 * Mirrors the C# `Packet.Ax25.Session.GuardEvaluator` line-for-line.
 */
export type GuardBindings = ReadonlyMap<string, () => boolean>;

export class GuardEvaluationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GuardEvaluationError";
  }
}

export class GuardEvaluator {
  constructor(private readonly bindings: GuardBindings) {}

  evaluate(expression: string | null | undefined): boolean {
    if (expression == null) return true;
    const trimmed = expression.trim();
    if (trimmed === "") return true;
    const tokens = trimmed.split(/[ \t]+/).filter((t) => t.length > 0);
    const state = { tokens, idx: 0, original: expression };
    const result = this.parseOr(state);
    if (state.idx !== tokens.length) {
      throw new GuardEvaluationError(
        `trailing tokens in guard expression '${expression}' at position ${state.idx}`,
      );
    }
    return result;
  }

  private parseOr(state: ParseState): boolean {
    let result = this.parseAnd(state);
    while (state.idx < state.tokens.length && state.tokens[state.idx] === "or") {
      state.idx++;
      const right = this.parseAnd(state);
      result = result || right;
    }
    return result;
  }

  private parseAnd(state: ParseState): boolean {
    let result = this.parseFactor(state);
    while (
      state.idx < state.tokens.length &&
      state.tokens[state.idx] === "and"
    ) {
      state.idx++;
      const right = this.parseFactor(state);
      result = result && right;
    }
    return result;
  }

  private parseFactor(state: ParseState): boolean {
    if (state.idx >= state.tokens.length) {
      throw new GuardEvaluationError(
        `expected identifier in '${state.original}' at position ${state.idx}`,
      );
    }
    let negate = false;
    if (state.tokens[state.idx] === "not") {
      negate = true;
      state.idx++;
      if (state.idx >= state.tokens.length) {
        throw new GuardEvaluationError(
          `expected identifier after 'not' in '${state.original}'`,
        );
      }
    }
    const ident = state.tokens[state.idx++]!;
    const pred = this.bindings.get(ident);
    if (!pred) {
      throw new GuardEvaluationError(
        `unbound identifier '${ident}' in '${state.original}' — add a binding before evaluating this guard`,
      );
    }
    const value = pred();
    return negate ? !value : value;
  }
}

interface ParseState {
  readonly tokens: readonly string[];
  idx: number;
  readonly original: string;
}
